python/lsst/daf/butler/direct_query_driver/_result_page_converter.py

```diff
@@ -74,63 +83,137 @@ def convert(self, raw_rows: Iterable[sqlalchemy.Row], next_key: PageKey | None)
         raise NotImplementedError()
 
 
+@dataclass(frozen=True)
+class ResultPageConverterContext:
+    """Parameters used by all result page converters."""
+
+    db: Database
+    column_order: qt.ColumnOrder
+    dimension_record_cache: DimensionRecordCache
+
+
 class DimensionRecordResultPageConverter(ResultPageConverter):  # numpydoc ignore=PR01
     """Converts raw SQL rows into pages of `DimensionRecord` query results."""
 
-    def __init__(self, spec: DimensionRecordResultSpec, db: Database) -> None:
+    def __init__(self, spec: DimensionRecordResultSpec, ctx: ResultPageConverterContext) -> None:
         self._result_spec = spec
-        self._timespan_repr_cls = db.getTimespanRepresentation()
+        self._converter = _create_dimension_record_row_converter(spec.element, ctx, use_cache=False)
 
     def convert(
         self, raw_rows: Iterable[sqlalchemy.Row], next_key: PageKey | None
     ) -> DimensionRecordResultPage:
-        result_spec = self._result_spec
-        record_set = DimensionRecordSet(result_spec.element)
-        record_cls = result_spec.element.RecordClass
-        if isinstance(result_spec.element, SkyPixDimension):
-            pixelization = result_spec.element.pixelization
-            id_qualified_name = qt.ColumnSet.get_qualified_name(result_spec.element.name, None)
-            for raw_row in raw_rows:
-                pixel_id = raw_row._mapping[id_qualified_name]
-                record_set.add(record_cls(id=pixel_id, region=pixelization.pixel(pixel_id)))
-        else:
-            # Mapping from DimensionRecord attribute name to qualified column
-            # name, but as a list of tuples since we'd just iterate over items
-            # anyway.
-            column_map = list(
-                zip(
-                    result_spec.element.schema.dimensions.names,
-                    result_spec.element.dimensions.names,
-                )
+        record_set = DimensionRecordSet(self._result_spec.element)
+        for raw_row in raw_rows:
+            record_set.add(self._converter.convert(raw_row))
+        return DimensionRecordResultPage(spec=self._result_spec, next_key=next_key, rows=record_set)
+
+
+def _create_dimension_record_row_converter(
+    element: DimensionElement, ctx: ResultPageConverterContext, *, use_cache: bool = True
+) -> _DimensionRecordRowConverter:
+    if use_cache and element.is_cached:
+        return _CachedDimensionRecordRowConverter(element, ctx.dimension_record_cache)
+    elif isinstance(element, SkyPixDimension):
+        return _SkypixDimensionRecordRowConverter(element)
+    else:
+        return _NormalDimensionRecordRowConverter(element, ctx.db)
+
+
+class _DimensionRecordRowConverter:
```

Comment:
```suggestion
class _DimensionRecordRowConverter(ABC):
```
I think this is needed in order for `@abstractmethod` to do anything.

Reply:
Actually mypy will statically check it without the metaclass

https://mypy.readthedocs.io/en/stable/class_basics.html#abstract-base-classes-and-multiple-inheritance